File: java/src/main/java/ds/bst/BST.java
package ds.bst;

import java.util.ArrayList;
import java.util.List;

public class BST <T extends Comparable>{
    private Node root;
    public BST(T value) throws Exception {
        if(value==null){
            throw new Exception("Couldn't handle null value!");
        }
        root = new Node(value);
    }
    private class Node{
        private Node left;
        private Node right;
        T value;
        public Node(T value) {
            this.value = value;
        }
    }

    public void add(T value) throws Exception {
        if(value==null){
            throw new Exception("Couldn't handle null value!");
        }
        add(this.root,value);
    }
    private Node add(Node root,T value){
        if (root == null) {
            return new Node(value);
        }
        else if(root.value.compareTo(value)>0){
            root.left = add(root.left,value);
        }
        else if(root.value.compareTo(value)<0){
            root.right = add(root.right,value);
        }
        return root;
    }
    public T find(T value){
        return find(root,value);
    }
    private T find(Node root,T value){
        if(root == null){
            return null;
        }
        else if(root.value.compareTo(value)>0){
            return find(root.left,value);
        }
        else if(root.value.compareTo(value)<0){
            return find(root.right,value);
        }
        else {
            return root.value;
        }
    }
    public void delete(T value){
        delete(root, value);
    }
    private Node delete(Node root, T value){
        if(root == null){
            return null;
        }
        else if(root.value.compareTo(value)<0){
            root.right = delete(root.right, value);
        }
        else if(root.value.compareTo(value)>0){
            root.left = delete(root.left, value);
        }
        else{
            if(root.right == null){
                return root.left;
            }
            else if(root.left==null){
                return root.right;
            }
            else{
                Node p = root.left;
                if(p.right == null){
                    root.left = p.left;
                    root.value = p.value;
                }
                else{
                    while(p.right.right !=null)p=p.right;
                    root.value = p.right.value;
                    p.right = p.right.left;
                }
            }
        }
        return root;
    }
    public List<T> serialize(){
        return serialize(root);
    }
    private List<T> serialize(Node root){
        List<T> list = new ArrayList<T>();
        if(root != null){
            list.add(root.value);
            list.addAll(serialize(root.left));
            list.addAll(serialize(root.right));
        }
        return list;
    }
    public void deserialize(List<T> list){
        root = deserialize(list,0,list.size()-1);
    }
    private Node deserialize(List<T> list,int left,int right){
        if(left>right){
            return null;
        }
        else{
            T mid = list.get(left);
            Node root = new Node(mid);
            int i = left+1;
            while(i<=right && list.get(i).compareTo(mid)<0)i++;
            root.left = deserialize(list,left+1,i-1);
            root.right = deserialize(list,i,right);
            return root;
        }
    }
    public List<T> inOrder(){
        return midRoot(root);
    }
    private List<T> midRoot(Node root){
        List<T> list = new ArrayList<T>();
        if(root != null){
            list.addAll(midRoot(root.left));
            list.add(root.value);
            list.addAll(midRoot(root.right));
        }
        return list;
    }
    public static void main(String args[]) throws Exception {
        BST tree = new BST(10);
        int[] seq = {10,9,15,13,14,17};
        List<Integer> list = new ArrayList<Integer>();
        for(int i=0;i<seq.length;i++) {
            tree.add(seq[i]);
        }
        System.out.println(tree.serialize());
        tree.deserialize(tree.serialize());
        System.out.println(tree.serialize());
        System.out.println(tree.inOrder());
    }
}



/*
List<Node> tree;
*/

File: python/bst/BST.py
#encoding:utf-8
class Node:
    def __init__(self, value):
        self._value = value
        self._left = None
        self._right = None

class BST:
    _root=None
    def __init__(self,value):
        self._root=Node(value)

    def add(self,value):
        self.inner_add(value,self._root)

    def find(self,value):
        return self.inner_find(value,self._root)

    def delete(self,value):
        self.inner_delete(value,self._root)

    def serialize(self):
        return self.inner_serialize(self._root)

    def desirialize(self,list):
        self.inner_desirialize(self._root,list,0,len(list)-1)

    def inner_add(self, value, root):
        if root is None:
            return Node(value)
        elif root._value>value:
            root._left = self.inner_add(value, root._left)
        elif root._value<value:
            root._right = self.inner_add(value, root._right)
        return root

    def inner_find(self, value, root):
        if root is None:
            return False
        elif root._value>value:
            return self.inner_find(value, root._left)
        elif root._value<value:
            return self.inner_find(value, root._right)
        else:
            return True

    def inner_delete(self, value, root):
        if root is None:
            return None
        elif root._value>value:
            root._left = self.inner_delete(value, root._left)
        elif root._value<value:
            root._right = self.inner_delete(value, root._right)
        else:
            if root._left is None:
                return root._right
            elif root._right is None:
                return root._left
            else:
                p = root._left
                if p._right is None:
                    root._value = p._value
                    root._left = p._left
                else:
                    while not p._right._right is None:
                        p = p._right
                    root._value = p._right._value
                    p._right = p._right._left
        return root

    def inner_serialize(self,root):
        list = []
        if not root is None:
            list.append(root._value)
            list.extend(self.inner_serialize(root._left))
            list.extend(self.inner_serialize(root._right))
        return list

    def inner_desirialize(self,root,list,start,end):
        if start<=end:
            root._value = list[start]
            i = start +1
            while i<end and list[i]<root._value:
                i = i+1
            self.inner_desirialize(root._left,list,start+1,i-1)
            self.inner_desirialize(root._right,list,i,end)

if __name__ == '__main__':
    tree=BST(10)
    tree.add(15)
    tree.add(13)
    tree.add(14)
    tree.add(17)
    tree.add(9)

    print(tree.serialize())
    tree.desirialize(tree.serialize())
    print(tree.serialize())